### PR TITLE
fix: checks subtyping relation when decoding reference types in Candid

### PR DESCRIPTION
**Pull request by christoph-dfinity**
# Description

Implements Candid subtyping and applies it when decoding reference types, to get up to par with the current Candid spec.

At a user-level this means we will fail early when decoding something like:

`service { foo : (text) -> () }` at an expected type of `service { foo : (nat) -> () }`

rather than decoding successfully and only failing when actually trying to make the call to `foo` later on in the code.

Follow-up to https://github.com/dfinity/agent-js/pull/981

# How Has This Been Tested?

Additional unit tests

I also had to update one of the existing tests, as the binary value it was checking against actually had the wrong type

# Reviewer notes

Unfortunately there are a few formatting changes unrelated to the change itself. These were produced in 3bedee41f13a9027e79c4420fd379d1cd35be18c by running 
```
npm run -w packages/candid prettier:write
```
I added some tests, but it would probably be a good idea to write a parser for the test format in the spec and make sure this implementation conforms to it. I'd prefer doing that in another PR though. Up to you, dear reviewer, if we'd rather hold off on this change until then.

I added a global subtype cache that is reset before decoding a Candid message. As it's global and mutable, make sure to take a closer look if I messed up...

### CODE REVIEW

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,8 @@
 - fix: fixes a bug in the Ed25519KeyIdentity verify implementation where the argument order was incorrect
 - fix: fixes a bug in the `Principal` library where the management canister id util was incorrectly importing using `fromHex`
 - feat: change auth-client's default identity provider url
+- fix: Check subtyping relationship when decoding function or service references
```
Comment (krpeacock) [inline docs/CHANGELOG.md:20]: can you move this up to the unreleased section?

**Review by krpeacock — APPROVED** [review]
one comment on fixing the changelog, but approved

File: docs/CHANGELOG.md (hunk 1 above)
Comment (christoph-dfinity) [reply docs/CHANGELOG.md]: Done, thanks for noticing.

File: packages/candid/src/idl.test.ts [hunk 2]
```diff
@@ -928,3 +933,191 @@ describe('IDL opt edge cases (embedded)', () => {
...
+  });
+
+  describe('Subtyping on Vecs', () => {
+    testReflexive(IDL.Vec(IDL.Nat));
+    testSub(IDL.Vec(IDL.Nat), IDL.Vec(IDL.Int));
+    testSubFail(IDL.Vec(IDL.Int), IDL.Vec(IDL.Nat));
+  });
+
+  describe('Subtyping on Options', () => {
+    // Because of some of the more "special" rules around Option types
```
Comment (crusso) [inline packages/candid/src/idl.test.ts:988]: special indeed

File: packages/candid/src/idl.test.ts [hunk 3]
```diff
@@ -928,3 +933,191 @@ describe('IDL opt edge cases (embedded)', () => {
...
+
+    // Subtyping on individual fields
+    testSub(IDL.Record({ a: IDL.Nat }), IDL.Record({ a: IDL.Int }));
+    testSubFail(IDL.Record({ a: IDL.Int }), IDL.Record({ a: IDL.Nat }));
+
+    // Width subtyping
+    testSub(IDL.Record({ a: IDL.Nat, b: IDL.Nat }), IDL.Record({ a: IDL.Nat }));
+    testSubFail(IDL.Record({ a: IDL.Nat }), IDL.Record({ a: IDL.Nat, b: IDL.Nat }));
+
+    // Opt, Null, or Reserved fields are allowed to be missing
```
Comment (crusso) [inline packages/candid/src/idl.test.ts:1009]: Not just trailing ones, but in any position. Might be worth a couple more tests.

File: packages/candid/src/idl.ts [hunk 4]
```diff
@@ -2032,3 +2069,170 @@ export function Func(args: Type[], ret: Type[], annotations: string[] = []): Fun
...
+
+/**
+ * Subtyping on Candid types t1 <: t2 (Exported for testing)
+ * @param t1 The potential subtype
+ * @param t2 The potential supertype
+ */
+export function subtype(t1: Type, t2: Type): boolean {
+  const relations = subtypeCache.copy();
+  const isSubtype = subtype_(relations, t1, t2);
+  if (isSubtype) {
```
Comment (crusso) [inline packages/candid/src/idl.ts:2159]: Don't you want to remember the failed tests too?

File: packages/candid/src/idl.ts [hunk 5]
```diff
@@ -2032,3 +2069,170 @@ export function Func(args: Type[], ret: Type[], annotations: string[] = []): Fun
...
+    return new Relations(copy);
+  }
+
+  /// Returns whether we know for sure that a relation holds or doesn't (`true` or `false`), or
+  /// if we don't know yet (`undefined`)
+  known(t1: Type, t2: Type): boolean | undefined {
+    return this.rels.get(t1.name)?.get(t2.name);
+  }
+
+  addNegative(t1: Type, t2: Type) {
```
Comment (crusso) [inline packages/candid/src/idl.ts:2096]: This appears to be unused. Perhaps it should be use in subtype(t1,t2)?

**Review by crusso — COMMENTED** [review]
Instead of using a global cache, how would you feel about passing the relation argument as the intial argument to decodeValue, along with the other bit of state, the `Pipe`.
Then maybe type could have a  binary decodeValue(b, t) that dispatches to the decodeValue(relates, b, t) with a fresh, local cache?

Also, constructing an object to enumberate the fields of variants and records seems a little inefficient. Since the underlying _fields arrays are sorted by label hash, can you just traverse the arrays directly, without the extra allocation?

File: packages/candid/src/idl.test.ts [hunk 6]
```diff
@@ -928,3 +933,191 @@ describe('IDL opt edge cases (embedded)', () => {
...
+    const Even = IDL.Rec();
+    const Odd = IDL.Rec();
+    Even.fill(IDL.Tuple(Odd));
+    Odd.fill(IDL.Tuple(Even));
+
+    testSub(IDL.Tuple(Even), Odd);
+    testSub(IDL.Tuple(IDL.Tuple(Odd)), Odd);
+  });
+
+  describe('decoding function/service references', () => {
```
Comment (crusso) [inline packages/candid/src/idl.test.ts:1084]: Probably worth adding a test that a failing subtype test under an expected optional function/service type defaults to `null`.

**Review by crusso — COMMENTED** [review]
Very nice and clean implementation!

File: packages/candid/src/idl.ts [hunk 7]
```diff
@@ -1120,6 +1120,14 @@ export class RecordClass extends ConstructType<Record<string, any>> {
     return x;
   }
 
+  get fieldsAsObject(): Record<string, Type> {
```
Comment (crusso) [inline packages/candid/src/idl.ts:1123]: Convenient, but maybe inefficient too? Also, shouldn't we be comparing the fields by their hashes, not their names? If there's a collision, something might go subtly wrong.

File: packages/candid/src/idl.ts [hunk 8]
```diff
@@ -1315,6 +1323,14 @@ export class VariantClass extends ConstructType<Record<string, any>> {
     }
     throw new Error('Variant has no data: ' + x);
   }
+
+  get alternativesAsObject(): Record<string, Type> {
```
Comment (crusso) [inline packages/candid/src/idl.ts:1327]: Same again. Convenient, but maybe inefficient too? Also, shouldn't we be comparing the fields by their hashes, not their names? If there's a collision, something might go subtly wrong.

File: packages/candid/src/idl.test.ts (hunk 3 above)
Comment (christoph-dfinity) [reply packages/candid/src/idl.test.ts:1009]: As these tests are checking records, there shouldn't be any notion of trailing?

I did add a bunch of test cases for the function check.

**Comment by christoph-dfinity** [comment]
> Instead of using a global cache that needs to be reset carefully, how would you feel about passing the relation argument as the intial argument to decodeValue, along with the other bit of state, the Pipe.
Then maybe Type could have a binary decodeValue(b, t) that dispatches to the new ternary decodeValue(relations, b, t) with a fresh, local cache?

I got started on this by changing the signature of `Type.decodeValue` to include an optional argument for the cache (optional so I wouldn't cause a breaking change to the API), but found that after changing a few call sites, it was much easier to miss passing the cache in some place than to reset the global cache "carefully" in one place. Unfortunately you can't overload like you proposed in TypeScript.

> Also, constructing an object just to enumerate the fields of variants and records seems a little inefficient. Since the underlying _fields arrays (Array([string, Type]) appear to be sorted by hash of the string, can you just traverse the arrays directly, without the extra allocation? I guess that might mean recomputing he hashed before comparing the labels for relative order...

As discussed I decided to go for clarity over max performance in this case (matching the [Rust implementation](https://github.com/dfinity/candid/blob/30c388671462aecdc4a3a9753d50dc2e8208c200/rust/candid/src/types/subtype.rs#L81)). I think it's very unlikely to make a noticeable difference (especially with the subtyping cache around).

I think I addressed the rest of your feedback. PTAL

File: packages/candid/src/idl.ts (hunk 8 above)
Comment (crusso) [reply packages/candid/src/idl.ts:1327]: Perf can come later

File: packages/candid/src/idl.ts [hunk 9]
```diff
@@ -1878,6 +1914,7 @@ export function decode(retTypes: Type[], bytes: ArrayBuffer): JsonValue[] {
     }
   });
 
+  resetSubtypeCache();
```
Comment (crusso) [inline packages/candid/src/idl.ts:1917]: Is it worth resetting the cache once done, to avoid leaking space?

Comment (christoph-dfinity) [reply packages/candid/src/idl.ts:1917]: good call, done in c540f2522985aa685915f89abde2ac35254f6e8e

File: packages/candid/src/idl.ts [hunk 10]
```diff
@@ -2032,3 +2069,174 @@ export function Func(args: Type[], ret: Type[], annotations: string[] = []): Fun
...
+    return this.rels.get(t1.name)?.get(t2.name);
+  }
+
+  addNegative(t1: Type, t2: Type) {
+    this.addNames(t1.name, t2.name, false);
+  }
+
+  add(t1: Type, t2: Type) {
+    this.addNames(t1.name, t2.name, true);
+  }
```
Comment (crusso) [inline packages/candid/src/idl.ts:2108]: Maybe we shouldn't worry but t.name is a property that is recomputed on each access.  It's quite expensive for record and variants. Do we want to address this, or leave it as an issue? It's a shame we can't use something else to key on.

**Comment by christoph-dfinity** [comment]
I'll set up some benchmarks before trying to think too hard about what's fast or slow here... Thanks for the review I think we're good to merge for now.

**Comment by christoph-dfinity** [comment]
@krpeacock Looks like I don't have the permissions to merge myself. Would you do me the honor?
